File: cocos/core/pipeline/ui/ui-flow.ts
/**
 * @category pipeline
 */
import { ccclass } from '../../data/class-decorator';
import { ForwardFlowPriority } from '../forward/enum';
import { IRenderFlowInfo, RenderFlow } from '../render-flow';
import { RenderView } from '../render-view';
import { UIStage } from './ui-stage';
import { RenderFlowTag } from '../pipeline-serialization';
import { sceneCulling } from '../forward/scene-culling';
import { ForwardPipeline } from '../forward/forward-pipeline';

/**
 * @en The UI render flow
 * @zh UI渲染流程。
 */
@ccclass('UIFlow')
export class UIFlow extends RenderFlow {

    public static initInfo: IRenderFlowInfo = {
        name: 'UIFlow',
        priority: ForwardFlowPriority.UI,
        tag: RenderFlowTag.UI,
    };

    public initialize (info: IRenderFlowInfo): boolean {

        super.initialize(info);

        const uiStage = new UIStage();
        uiStage.initialize(UIStage.initInfo);
        this._stages.push(uiStage);

        return true;
    }

    public destroy () {
        super.destroy();
    }

    public render (view: RenderView) {
        const pipeline = this._pipeline as ForwardPipeline;
        view.camera.update();
        sceneCulling(pipeline, view);
        pipeline.updateUBOs(view);
        super.render(view);
    }
}

File: native/cocos/renderer/pipeline/ui/UIFlow.cpp

#include "UIFlow.h"
#include "../Define.h"
#include "../forward/ForwardPipeline.h"
#include "../forward/SceneCulling.h"
#include "UIStage.h"

namespace cc {
namespace pipeline {
RenderFlowInfo UIFlow::_initInfo = {
    "UIFlow",
    static_cast<uint>(ForwardFlowPriority::UI),
    static_cast<uint>(RenderFlowTag::UI),
    {}
};
const RenderFlowInfo &UIFlow::getInitializeInfo() { return UIFlow::_initInfo; }

UIFlow::~UIFlow() {
}

bool UIFlow::initialize(const RenderFlowInfo &info) {
    RenderFlow::initialize(info);
    if (_stages.size() == 0) {
        auto uiStage = CC_NEW(UIStage);
        uiStage->initialize(UIStage::getInitializeInfo());
        _stages.emplace_back(uiStage);
    }
    return true;
}

void UIFlow::activate(RenderPipeline *pipeline) {
    RenderFlow::activate(pipeline);
}

void UIFlow::render(RenderView *view) {
    auto pipeline = static_cast<ForwardPipeline *>(_pipeline);
    pipeline->updateUBOs(view);
    RenderFlow::render(view);
}

void UIFlow::destroy() {
    RenderFlow::destroy();
}

} // namespace pipeline
} // namespace cc
